{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "386aa57f-8b67-461d-9a25-e97fcc8b119f",
   "metadata": {},
   "source": [
    "# 1. Set up packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df0cbd9e-1085-47bf-92ee-f46ce1a6829f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "import subprocess\n",
    "\n",
    "# Helper to run shell commands\n",
    "def run(cmd, check=True):\n",
    "    print(f\"$ {cmd}\")\n",
    "    with open(\"output.log\", \"w\") as out_file, open(\"error.log\", \"w\") as err_file:\n",
    "        proc = subprocess.run(cmd, shell=True, stdout=out_file, stderr=err_file, text=True)\n",
    "        print(proc.stdout)\n",
    "        if proc.returncode != 0:\n",
    "            print(proc.stderr)\n",
    "            if check:\n",
    "                raise RuntimeError(f\"Command failed: {cmd}\")\n",
    "    return proc\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5799e751-9b31-49f1-9dea-b241c06149bb",
   "metadata": {},
   "source": [
    "# 2. Download and compress fastq files from two timepoints from SRA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c220bbf6-b4fd-4048-b8ca-ee108fe8dd7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download fastq files from SRA using fasterq-dump\n",
    "run(\"fasterq-dump --split-files SRR13973794\")\n",
    "run(\"fasterq-dump --split-files SRR13973795\")\n",
    "\n",
    "# Compress fastq files to reduce file size\n",
    "run(\"gzip SRR1397379*.fastq\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e7ef7d-b461-4d0c-93c2-835788e3b282",
   "metadata": {},
   "source": [
    "# 3. Run FastQC to check sequence quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15164402-6a77-4c68-a194-f19f43a9a7ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run FastQC on raw reads\n",
    "run(\"mkdir fastqc_raw\")\n",
    "run(\"fastqc -o fastqc_raw SRR13973794_1.fastq.gz SRR13973794_2.fastq.gz\")\n",
    "run(\"fastqc -o fastqc_raw SRR13973795_1.fastq.gz SRR13973795_2.fastq.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f6c8ef3-d75a-442d-bf57-4106dc7d6a92",
   "metadata": {},
   "source": [
    "# 4. Trimming - fastp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2474ce5-b05d-4ae4-8434-cd025c122e22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Trimming adapter and short sequences using fastp\n",
    "run(\"fastp -i SRR13973794_1.fastq.gz \\\n",
    "    -I SRR13973794_2.fastq.gz \\\n",
    "    -o SRR13973794_1.trim.fastq.gz \\\n",
    "    -O SRR13973794_2.trim.fastq.gz \\\n",
    "    --detect_adapter_for_pe \\\n",
    "    --trim_poly_g \\\n",
    "    --thread 8 \\\n",
    "    -j SRR13973794_fastp.json -h SRR13973794_fastp.html\")\n",
    "\n",
    "run(\"fastp -i SRR13973795_1.fastq.gz \\\n",
    "    -I SRR13973795_2.fastq.gz \\\n",
    "    -o SRR13973795_1.trim.fastq.gz \\\n",
    "    -O SRR13973795_2.trim.fastq.gz \\\n",
    "    --detect_adapter_for_pe \\\n",
    "    --trim_poly_g \\\n",
    "    --thread 8 \\\n",
    "    -j SRR13973795_fastp.json -h SRR13973795_fastp.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aca1a6c0-4a89-49a0-aa48-2867831b472a",
   "metadata": {},
   "source": [
    "# 5. Rerun FastQC after trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3179aa-2fe2-4460-ac46-2a9f0218dbfa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 5. Rerun FastQC after trimming to confirm quality\n",
    "run(\"mkdir fastqc_trimmed\")\n",
    "run(\"fastqc -o fastqc_trimmed SRR13973794_1.trim.fastq.gz SRR13973794_2.trim.fastq.gz\")\n",
    "run(\"fastqc -o fastqc_trimmed SRR13973795_1.trim.fastq.gz SRR13973795_2.trim.fastq.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e8417c8-661f-4ebf-b424-d8838da2b17e",
   "metadata": {},
   "source": [
    "# 6. Reference preparation - bowtie2, samtools, picard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c773d8c4-385e-4db2-8427-4aea77967fd3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Indexing reference genome and creating .fai and .dict files that will be needed later\n",
    "### Only need to run once\n",
    "#run(\"bowtie2-build Homo_sapiens_assembly38.fasta Homo_sapiens_assembly38.fasta --threads 8\")\n",
    "#run(\"samtools faidx Homo_sapiens_assembly38.fasta\")\n",
    "#run(\"picard CreateSequenceDictionary R=Homo_sapiens_assembly38.fasta O=Homo_sapiens_assembly38.dict\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b4bf174-7990-4736-aa31-3d351f7536dd",
   "metadata": {},
   "source": [
    "# 7. Alignment - bowtie2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b88f94f-d71f-4d2c-993b-c2d8d47877cc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Run bowtie 2 to align reads for both samples\n",
    "run('bowtie2 --rg-id plasma_99 \\\n",
    "--rg SM:plasma_99 \\\n",
    "--rg PL:ILLUMINA \\\n",
    "--rg LB:Targeted-Capture \\\n",
    "--rg PU:HiSeq2500 \\\n",
    "--rg DS:\"Strategy=Targeted-Capture;Source=GENOMIC;Selection=Hybrid-Selection\" \\\n",
    "-p 8 \\\n",
    "-x Homo_sapiens_assembly38.fasta \\\n",
    "-1 SRR13973794_1.trim.fastq.gz \\\n",
    "-2 SRR13973794_2.trim.fastq.gz \\\n",
    "-S SRR13973794_aligned_reads.sam')\n",
    "\n",
    "run('bowtie2 --rg-id plasma_98 \\\n",
    "--rg SM:plasma_98 \\\n",
    "--rg PL:ILLUMINA \\\n",
    "--rg LB:Targeted-Capture \\\n",
    "--rg PU:HiSeq2500 \\\n",
    "--rg DS:\"Strategy=Targeted-Capture;Source=GENOMIC;Selection=Hybrid-Selection\" \\\n",
    "-p 8 \\\n",
    "-x Homo_sapiens_assembly38.fasta \\\n",
    "-1 SRR13973795_1.trim.fastq.gz \\\n",
    "-2 SRR13973795_2.trim.fastq.gz \\\n",
    "-S SRR13973795_aligned_reads.sam')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de7b65ec-516c-4264-997c-1dc1a18001c9",
   "metadata": {},
   "source": [
    "# 8. Convert SAM files to BAM - samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60b960c8-31fc-4fcb-8628-7f89caa9f5e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert aligned SAM files to BAM\n",
    "run(\"samtools view -bS SRR13973794_aligned_reads.sam > SRR13973794_aligned_reads.bam\")\n",
    "run(\"samtools view -bS SRR13973795_aligned_reads.sam > SRR13973795_aligned_reads.bam\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17bfd723-2693-4333-acc4-f5ddb8a08df2",
   "metadata": {},
   "source": [
    "# 9. Extract APC reads - samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e845b700-7e1b-4390-b2f9-649a78a0c181",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract only reads that overlap APC gene\n",
    "\n",
    "# Using grep to get coordinates for GTF file\n",
    "run(\"grep ENSG00000134982 gencode.v47.basic.protein_coding.canonical.gtf\") # APC gene coords. = chr5:112707518-112846239\n",
    "\n",
    "# Using samtools to extract reads from APC\n",
    "# Sort reads first\n",
    "run(\"samtools sort -o SRR13973794_aligned_reads.sorted.bam SRR13973794_aligned_reads.bam\")\n",
    "run(\"samtools sort -o SRR13973795_aligned_reads.sorted.bam SRR13973795_aligned_reads.bam\")\n",
    "\n",
    "# index sorted bam file\n",
    "run(\"samtools index SRR13973794_aligned_reads.sorted.bam\")\n",
    "run(\"samtools index SRR13973795_aligned_reads.sorted.bam\")\n",
    "\n",
    "# Select only reads overlaping the APC gene location\n",
    "run(\"samtools view -b SRR13973794_aligned_reads.sorted.bam chr5:112707518-112846239 > SRR13973794_APC_reads.bam\")\n",
    "run(\"samtools view -b SRR13973795_aligned_reads.sorted.bam chr5:112707518-112846239 > SRR13973795_APC_reads.bam\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96252cc9-aad8-4bbf-a1e5-1f235d5eab7a",
   "metadata": {},
   "source": [
    "# 10. Mark duplicates - MarkDuplicatesSpark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59ca2fd2-56ac-4a6e-9e7b-6e4fdc38b332",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Mark duplicates in aligned read BAM files\n",
    "run(\"gatk MarkDuplicatesSpark \\\n",
    "    -I SRR13973794_APC_reads.bam \\\n",
    "    -O SRR13973794_APC_marked_duplicates.bam \\\n",
    "    -M SRR13973794_APC_marked_dup_metrics.txt\")\n",
    "    \n",
    "run(\"gatk MarkDuplicatesSpark \\\n",
    "    -I SRR13973795_APC_reads.bam \\\n",
    "    -O SRR13973795_APC_marked_duplicates.bam \\\n",
    "    -M SRR13973795_APC_marked_dup_metrics.txt\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87a4be6a-1111-4e38-947e-fd41f043593d",
   "metadata": {},
   "source": [
    "# 11. Base score recalibration - GATK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5035c72-d0d8-4d10-9b55-cf191e926908",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use GATK for base score recalibration based on known hg38 indels sites\n",
    "run(\"gatk BaseRecalibrator \\\n",
    "  -I SRR13973794_APC_marked_duplicates.bam \\\n",
    "  -R Homo_sapiens_assembly38.fasta \\\n",
    "  --known-sites Mills_and_1000G_gold_standard.indels.hg38.vcf.gz \\\n",
    "  -O SRR13973794_APC_marked_duplicates_recal_data.table\")\n",
    "  \n",
    "run(\"gatk ApplyBQSR \\\n",
    "    -R Homo_sapiens_assembly38.fasta \\\n",
    "    -I SRR13973794_APC_marked_duplicates.bam \\\n",
    "    --bqsr-recal-file SRR13973794_APC_marked_duplicates_recal_data.table \\\n",
    "    -O SRR13973794_APC_marked_duplicates_recal_data.bam\")\n",
    "\n",
    "run(\"gatk BaseRecalibrator \\\n",
    "  -I SRR13973795_APC_marked_duplicates.bam \\\n",
    "  -R Homo_sapiens_assembly38.fasta \\\n",
    "  --known-sites Mills_and_1000G_gold_standard.indels.hg38.vcf.gz \\\n",
    "  -O SRR13973795_APC_aligned_reads_recal_data.table\")\n",
    "  \n",
    "run(\"gatk ApplyBQSR \\\n",
    "    -R Homo_sapiens_assembly38.fasta \\\n",
    "    -I SRR13973795_APC_marked_duplicates.bam \\\n",
    "    --bqsr-recal-file SRR13973795_aligned_reads_recal_data.table \\\n",
    "    -O SRR13973795_APC_marked_duplicates_recal_data.bam\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79947892-df75-410c-8216-4eebad53fb81",
   "metadata": {},
   "source": [
    "# 12. Create mpilup files - samtools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1d243b0-63db-444a-93a5-949b6e0ab39d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use samtools to convert BAM files to mpileup format\n",
    "run(\"samtools mpileup -f Homo_sapiens_assembly38.fasta SRR13973794_APC_marked_duplicates_recal_data.bam > SRR13973794_APC.pileup\")\n",
    "run(\"samtools mpileup -f Homo_sapiens_assembly38.fasta SRR13973795_APC_marked_duplicates_recal_data.bam > SRR13973795_APC.pileup\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e98883df-16be-4880-a1df-c37861a95b50",
   "metadata": {},
   "source": [
    "# 13. Call variants - VarScan2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a6d2d7e-20ec-4d63-800c-02e3ba583762",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Run VarScan2 to call SNP variants\n",
    "run(\"java -jar ./VarScan.v2.3.9.jar mpileup2snp SRR13973794_APC.pileup --min-coverage 10 --min-avg-qual 20 --min-var-freq 0.20 --p-value 0.05 --output-vcf 1 > SRR13973794_APC.snp.vcf\")\n",
    "run(\"java -jar ./VarScan.v2.3.9.jar mpileup2snp SRR13973795_APC.pileup --min-coverage 10 --min-avg-qual 20 --min-var-freq 0.20 --p-value 0.05 --output-vcf 1 > SRR13973795_APC.snp.vcf\")\n",
    "\n",
    "# Compare pre and post treatment\n",
    "run(\"ava -jar VarScan.v2.3.9.jar somatic SRR13973794_APC.pileup SRR13973795_APC.pileup --min-coverage 10 --min-var-freq 0.20 --p-value 0.05 --somatic-p-value 0.05 --output-vcf 1 > comp_APC.snp.vcf\")\n",
    "\n",
    "# 17 SNPs in pre-treatment samples and 21 SNPs in during treatment sample, shows that there are new somatic mutations that have appeared\n",
    "# and the tumor is likely evolving and potentially developing resistance to the current treatment\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b97a848-ca66-4fae-8cf1-6d00a305d9d5",
   "metadata": {},
   "source": [
    "# 14. Annotate variants - snpEff and snpSift"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5be9f7e-f63a-4854-991b-49a057fd09c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### NOTE: Need JAVA 21+ to be able to run snpEff and snpSift\n",
    "# Use snpEff to annotate the gene variants\n",
    "run(\"java -Xmx4G -jar ./snpEff/snpEff.jar eff -v GRCh38.86 SRR13973794_APC.snp.vcf > pre_treatment_APC_calls.filtered.vcf\")\n",
    "run(\"java -Xmx4G -jar ./snpEff/snpEff.jar eff -v GRCh38.86 SRR13973795_APC.snp.vcf > during_treatment_APC_calls.filtered.vcf\")\n",
    "\n",
    "# Use snpSift to filter for only variants that passed snpEff filter and add ClinVAR designations to the annotated gene variants\n",
    "!java -jar ./snpEff/SnpSift.jar filter \"(FILTER = 'PASS')\" pre_treatment_APC_calls.filtered.vcf |java -jar ./snpEff/SnpSift.jar annotate clinvar_20251109.vcf.gz > pre_treatment_APC_annotated.vcf\n",
    "!java -jar ./snpEff/SnpSift.jar filter \"(FILTER = 'PASS')\" post_treatment_APC_calls.filtered.vcf |java -jar ./snpEff/SnpSift.jar annotate clinvar_20251109.vcf.gz > during_treatment_APC_annotated.vcf\n",
    "\n",
    "## Do not see any previously described clinical variants in the APC gene for this patient. Could look at other clinical annotation\n",
    "## datasets such as COSMIC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dbc6e6b-ac3d-49d6-bce7-21ed9a53343b",
   "metadata": {},
   "source": [
    "# 15. Next steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92572155-ce3d-4942-8561-5a2d7ee3bf5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create visualization of APC gene with locations of the SNPs\n",
    "# Run all samples and see if trend towards increased SNPs in APC holds true and if SNPs occure in the same places\n",
    "# Create a simple plot of APC SNPs before and during/after treatment with anti-EGFR drugs"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
